prusti-viper/src/encoder/encoder.rs

```diff
@@ -249,10 +259,248 @@ impl<'v, 'tcx> Encoder<'v, 'tcx> {
                 type_vars: vec![],
             });
         }
+
+        if config::enable_purification_optimization() {
+            domains.push(self.axiomatized_function_domain.borrow().clone());
+            domains.push(self.get_nat_domain());
+            domains.push(self.get_primitive_valid_domain())
+        }
+
         domains.sort_by_key(|d| d.get_identifier());
         domains
     }
 
+    fn get_primitive_valid_domain(&self) -> vir::Domain {
+        //FIXME this does not check or handel the different sizes of primitve types
+        let domain_name = "PrimitiveValidDomain";
+
+        let mut functions = vec![];
+        let mut axioms = vec![];
+        for t in &[vir::Type::Bool , vir::Type::Int] {
+            let f = snapshot::valid_func_for_type(t);
+            functions.push(f.clone());
+
+            let forall_arg  = vir::LocalVar{name: "self".to_owned(), typ: t.clone()};
+            let function_app = vir::Expr::domain_func_app(f.clone(), vec![vir::Expr::local(forall_arg.clone())]);
+            let body = vir::Expr::and(function_app, true.into());
+            let e = vir::Expr::forall(vec![forall_arg], vec![], body); //TODO triggers
+            let ax = vir::DomainAxiom{  name: format!("{}$axiom", f.get_identifier()),
+                expr: e,
+                domain_name: domain_name.to_string()};
+            axioms.push(ax); //TODO
+        }
+
+
+        vir::Domain {
+            name: domain_name.to_owned(),
+            functions,
+            axioms,
+            type_vars: vec![],
+        }
+    }
+
+    fn get_succ_func(&self) -> vir::DomainFunc {
+        let succ = vir::DomainFunc {
+            name: "succ".to_owned(),
+            formal_args: vec![vir::LocalVar {
+                name: "val".to_owned(),
+                typ: vir::Type::Domain(snapshot::NAT_DOMAIN_NAME.to_owned()),
+            }],
+            return_type: vir::Type::Domain(snapshot::NAT_DOMAIN_NAME.to_owned()),
+            unique: false,
+            domain_name: snapshot::NAT_DOMAIN_NAME.to_owned(),
+        };
+
+        succ
+    }
+    fn get_nat_domain(&self) -> vir::Domain {
+        let nat_domain_name = snapshot::NAT_DOMAIN_NAME;
+        let zero = vir::DomainFunc {
+            name: "zero".to_owned(),
+            formal_args: vec![],
+            return_type: vir::Type::Domain(nat_domain_name.to_owned()),
+            unique: false,
+            domain_name: nat_domain_name.to_owned(),
+        };
+       
+        let functions = vec![zero, self.get_succ_func()];
+
+        vir::Domain {
+            name: nat_domain_name.to_owned(),
+            functions,
+            axioms: vec![],
+            type_vars: vec![],
+        }
+    }
+
+    fn encode_axiomatized_pure_function(&self, f: &vir::Function) {
+        let snapshots_info: HashMap<String, Box<Snapshot>> = self.snapshots.borrow().clone();
```

Comment:
I wonder if it's possible to avoid cloning the map. The `ExprPurifier` could take a reference to the map, for example.

Reply:
That was in fact possible. Thank you for pointing this out.